File: strings/determine-if-string-halves-are-alike/determine-if-string-halves-are-alike.go
package determineifstringhalvesarealike

import "unicode"

func halvesAreAlike(s string) bool {
	n := len(s)

	firstCount := 0
	secondCount := 0

	i := 0
	j := n / 2

	for j < n {
		if isVowel(rune(s[i])) {
			firstCount++
		}

		if isVowel(rune(s[j])) {
			secondCount++
		}

		i++
		j++
	}

	return firstCount == secondCount

}

func isVowel(c rune) bool {
	switch unicode.ToLower(c) {
	case 'a':
		return true
	case 'e':
		return true
	case 'i':
		return true
	case 'o':
		return true
	case 'u':
		return true
	}
	return false
}

File: strings/determine-if-string-halves-are-alike/determine-if-string-halves-are-alike.java
class Solution {
  public boolean halvesAreAlike(String s) {
    int n = s.length();

    int firstCount = 0;
    int secondCount = 0;

    int i = 0;
    int j = n / 2;

    while (j < n) {
      if (isVowel(s.charAt(i))) {
        firstCount++;
      }

      if (isVowel(s.charAt(j))) {
        secondCount++;
      }

      i++;
      j++;
    }

    return firstCount == secondCount;
  }

  private boolean isVowel(char c) {
    switch (Character.toLowerCase(c)) {
    case 'a':
      return true;
    case 'e':
      return true;
    case 'i':
      return true;
    case 'o':
      return true;
    case 'u':
      return true;
    default:
      return false;
    }
  }
}
